### ISSUE TITLE
`dart analyze` should use LSP to communicate with analysis server

### ISSUE THREAD

**Issue by srawlins** [dart-lang/sdk#55075]
Right now it uses the legacy protocol, but that's legacy...

**Comment by DanTup** [comment]
Slightly related - in https://github.com/dart-lang/sdk/issues/54712 we were talking about debouncing analysis status events. It was noted that `dart analyze` uses these (so we should be sure any debouncing doesn't affect this), but that's only the legacy protocol and therefore debouncing in LSP would have no impact.. that may not be true if we move this over.

So, if/when we do this, we need to ensure there's a reliable way to know analysis is complete (and using the same status events used to control the indicator in an LSP client may or may not be the right thing to use).

**Comment by DanTup** [comment]
> So, if/when we do this, we need to ensure there's a reliable way to know analysis is complete (and using the same status events used to control the indicator in an LSP client may or may not be the right thing to use).

I'm currently working on a new request for this (https://github.com/dart-lang/sdk/issues/63772), so it might be a good time to move this over once that's done.

I'll tentatively assign this to me assuming it'll be a simple 😄

**Comment by bwilkerson** [comment]
Note that the `flutter analyze` command has already been converted to use LSP. I talked to Ben and there doesn't appear to be any good way to share the code, nor do I expect you'll need to look at the code, but I'm just trying to spread information in case it's useful in some way I haven't anticipated.

**Comment by DanTup** [comment]
Yeah, I did wonder about that. I think it might be slightly simpler here in that I think we can use `package:language_server_protocol` classes here (but I presume `flutter` could not). It would be nice to avoid too much duplication though - I'll look at the Flutter version before starting anything for this and will ping you both if I have any useful ideas.

**Comment by bwilkerson** [comment]
> ... we can use `package:language_server_protocol` classes here ...

True.

> ... (but I presume flutter could not).

My understanding is that this is also true. I believe that Flutter can only depend on published packages. I'm hesitant to public `package:language_server_protocol`, or anything derived from it [1] but maybe we should consider that.

[1] My original idea was to build a minimal package similar to what `analysis_servrer_client` used to be. The package would support the protocols that our command line tools need and nothing more. But I wouldn't want to publish such a package unless it supported the whole protocol.

**Comment by DanTup** [comment]
@bwilkerson I started looking at this and made some discoveries.

1. `dart fix` also uses the same `AnalysisServer` class that wraps the server, so we need to either migrate both at the same time, or create an `LspAnalysisServer` class alongside it and then we can work on them individually (I'm currently looking at doing the latter because it'll result in smaller reviews, and I'm not yet certain if LSP has everything needed for `dart fix`)
2. In the machine + json outputs of `dart analyze` we output the `type` field from `AnalysisError`, but we don't actually map this to anything in LSP so we don't have this information anymore (we have severity and code, but not the `AnalysisErrorType` (CHECKED_MODE_COMPILE_TIME_ERROR/COMPILE_TIME_ERROR/HINT/LINT/TODO/etc..))
3. In the json output of errors, we currently output offset+line+col but in LSP we only have line+col. We'd need to read the content and get a LineInfo to get the offset, but that conversion is risky without being certain the content hasn't changed
4. In the json output of errors, we have a `correctionMessage` field, but for LSP this got concatenated with the message

Any suggestions on how we should handle these? Some initial thoughts (which I may not have entirely thought through):

- use a custom request to get diagnostics with the original data - though that's a bit of a reversal from our plan for https://github.com/dart-lang/sdk/issues/63772
- assume content on disk hasn't changed to get LineInfos, handle invalid locations, try to look up AnalysisErrorTypes from codes - we might be able to get most of the way to what we had before, but this feels messy
- drop the data we don't have and announce it as a future breaking change
- or similarly, implement behind a flag (`dart analyze --protocol=lsp`) and keep both for now, so we can choose when to make that break more carefully

**Comment by bwilkerson** [comment]
> 1. `dart fix` also uses the same `AnalysisServer` class that wraps the server ...

Yes, it does. I assume that `dart migrate` also uses that class, though it operates via a custom LSP protocol, which I'm assuming we're sending wapped over the legacy channel.

I'm fairly certain that LSP doesn't have everything we need for `dart fix` and that we'll need to add a custom protocol for it.

My assumption was that we'd convert `dart analyze` first, as a new class, and then later add the custom protocol and convert `dart fix` to use it. And if we have the new class soon enough, we can use it to implement the new `dart refactor` command (or whatever it ends up being called).

> 2. In the machine + json outputs of `dart analyze` we output the `type` field from `AnalysisError` ...

Ouch. I have no idea whether anyone is using those values. I suppose we'd need to send out a breaking change notice to see whether anyone really needs them. I assume that we could use custom fields on a `Diagnostic`, maybe behind a command-line flag that only `dart analyze` passes in so that we don't increase the amount of data being sent for normal operation. But hopefully nobody is depending on the `type`.

@srawlins Do you know whether these values are needed?

> 3. In the json output of errors, we currently output offset+line+col but in LSP we only have line+col.

I think it's safe to assume that nothing is changing on disk while `dart analyze` is running, so we can compute the offset in the `dart analyze` process. I think most users don't expect to be able to change what's on disk while a command-line tool like this is running.

> 4. In the json output of errors, we have a correctionMessage field, but for LSP this got concatenated with the message

Not having that separation will limit our formatting options, but I think we can just live with it.

I'm guessing that the machine + json outputs also have these as separate fields, so we'll need to include that in the breaking change notice.

> use a custom request to get diagnostics with the original data

I'd prefer to avoid that if we reasonably can, even though I kind of suggested the same thing above (adding a custom field). :-/

> ... implement behind a flag ...

I'd rather avoid that too, if possible. Seems like it would be better to just make the breaking change now than to postpone the inevitable.

**Comment by DanTup** [comment]
> Yes, it does. I assume that `dart migrate` also uses that class, though it operates via a custom LSP protocol, which I'm assuming we're sending wapped over the legacy channel.

I'm failing to find an implementation for this - is it still WIP? (I would've expected to find a command in https://github.com/dart-lang/sdk/tree/main/pkg/dartdev/lib/src/commands ?)

(Edit: I found https://github.com/dart-lang/sdk/issues/63765 (via @FMorschel) which is still open, so I presume only the server-side parts have landed so far)


> My assumption was that we'd convert `dart analyze` first, as a new class, and then later add the custom protocol and convert `dart fix` to use it.

Great, this is how I've started :) It would've been nice if the diff was simpler (eg. modifying the existing class so you can just see the protocol changes) rather than it being a whole new implementation to review, but perhaps this is better because it also means I can change the API where things work a bit differently.


> I think it's safe to assume that nothing is changing on disk while `dart analyze` is running, so we can compute the offset in the `dart analyze` process. I think most users don't expect to be able to change what's on disk while a command-line tool like this is running.

It does mean if there are a lot of diagnostics, we may read a lot of files (again) from disk and compute their `LineInfo`s (including for context messages), though hopefully that's not the most common case.


> Seems like it would be better to just make the breaking change now than to postpone the inevitable.

Yeah, sounds fair. I'll try and get things working fully and compute the things we can (like offsets) to compile a list of exactly what we'll lose, then we can review and compile a breaking change notice. Thanks!

**Comment by FMorschel** [comment]
> I'm failing to find an implementation for this

Related issue: https://github.com/dart-lang/sdk/issues/63240

**Comment by bwilkerson** [comment]
If there's no command yet, then I assume that @kallentu  is implementing the server side first. You can find the handler here: `pkg/analysis_server/lib/src/lsp/handlers/custom/handler_migrate.dart`.

**Comment by DanTup** [comment]
>  I assume that we could use custom fields on a Diagnostic

I have the `dart analyze` change mostly working and was making a note of what things change as a result of not having all the same fields, I found that `Diagnostic` has a `data` field. The intention is for it be round-tripped for codeAction requests (though we don't use diagnostics sent in code action requests at all).

So I think maybe rather than breaking changes, we should have a flag (probably an custom client capability) and include the data here. The client code can just ensure it's a `Map` before it reads it, and then look for known fields (`int? offset`, `String type`) and then we could preserve existing behaviour rather than having breaking changes (or at least, we could decouple those breaks from this migration).

The only slightly odd one might be `correctionMessage`. Usually we include that tagged onto the message. If we instead include it here in `data` so we can preserve having it in a separate field for the json output, do we:

a) also include `problemMessage` in `data` (without the correctionMessage appended)
b) not append it to the normal diagnostics message string if we are including it in `data`?

@bwilkerson any preferences (or other opinions on this)?

**Comment by DanTup** [comment]
> @bwilkerson any preferences (or other opinions on this)?

I prepared a CL at https://dart-review.googlesource.com/c/sdk/+/528080 with what I think might be the best option here, if seeing the code changes is easier to understand :-)

**Comment by bwilkerson** [comment]
> a) also include `problemMessage` in `data` (without the correctionMessage appended)

I don't think there's any need to send it twice. Given that the problem message is available in the normal protocol, I'd say that covers the requirements.

> b) not append it to the normal diagnostics message string if we are including it in `data?`

I think that makes sense. Otherwise clients would need to truncate the message string when there's a correction message, which is work that we don't need to make them do.

> I prepared a CL ...

That looks good to me. Are you ready for it to be reviewed?

**Comment by DanTup** [comment]
> That looks good to me. Are you ready for it to be reviewed?

Yep, I just wanted to check it seemed reasonable. I've added you as reviewer - thanks!

**Comment by DanTup** [comment]
I forgot to tag this, it landed in https://dart-review.googlesource.com/c/sdk/+/534060 / f0c0ab967e9d3f85d73b3fcd13da29a0277daa01

Bots are green, but please do let me know if you see any weirdness in this area that might be related!

(`dart fix` is still outstanding, the first step for that is ensuring LSP has everything it requires - https://github.com/dart-lang/sdk/issues/64013)
